File: council/polymorphism.rb
class HighCouncil::Factory
  def self.build(member)
    factory = new(member)
    factory.presenter.new(member)
  end

  def initialize(member)
    @member = member
  end

  def presenter
    HighCouncil.const_get("Presenter::#{member.council_status.camelcase}")
  end
end

# This is our type field
# jedi.council_status
# => 'sith in disguise'
# => 'yoda'
# => 'jedi'

class HighCouncil::Presenter::SithInDisguise
end

class HighCouncil::Presenter::Guest
end

class HighCouncil::Presenter::Yoda
end

class HighCouncil::Presenter::Jedi
  attr_reader :jedi

  def initialize(jedi)
    @jedi = jedi
  end

  def method_missing(sym, *args, &block)
    @jedi.send(sym, args, block)
  end
end

File: csharp/council/polymorphism.cs
"controller.cs"

class Jedi : ICouncilMember {
}

interface ICouncilMember {
  public string CouncilStatus()
}

namespace HighCouncil
{
  class Factory
  {
    public static ICouncilDecorator build(Jedi member)
    {
      factory = new Factory(member);
      Type type = factory.Decorator(member);
      Activator.CreateInstance(type, member);
    }

    public Factory(Jedi member)
    {
      member = member;
    }

    public Type Decorator()
    {
      // 'sith in disguise'
      // HighCouncil.Decorator.SithInDisguise
      Type type = Type.GetType("HighCouncil.Decorator." + member.CouncilStatus);
    }
  }

  namespace Decorator
  {
    class SithInDisguise : ICouncilDecorator {
    }

    class Guest : ICouncilDecorator {
    }

    class Yoda : ICouncilDecorator {
    }

    interface ICouncilDecorator {
    }

  }
}
